### ISSUE TITLE
build hooks

### ISSUE THREAD

**Issue by caarlos0** [goreleaser/goreleaser#1328]
check https://github.com/goreleaser/goreleaser/issues/1261 for details

it seems like something useful to have, for example, to run `upx` against the binaries 🤔

**Comment by radeksimko** [comment]
I feel this would also address #1411 

--- 

What do you think about new `each` and `all` configuration options under `after` - that would satisfy the current use case of 1 hook per build ID and also provide some flexibility where folks want to run hook for each binary

```yml
builds:
  -
    id: signable
    targets:
      - darwin_amd64
      - windows_386
      - windows_amd64
    after:
      each:
        - upx {{ .Path }}
      all:
        - zip ./signable_binaries.zip {{ range .Binaries }} {{.Path}}{{ end }}
        - ./codesign_binaries_in_zip -src=./signable_binaries.zip -dst=./signed_binaries.zip
        - unzip ./signed_binaries.zip
```
Technically these could also be combined with the assumption that `each` hooks run in parallel, right after the relevant build and `all` runs after all individual hooks are finished.

Would you be open for a PR that implements something like this? ^

**Comment by caarlos0** [comment]
fixed by #1414 right?

**Comment by radeksimko** [comment]
> fixed by #1414 right?

Yes.

--- 

I feel it would good to open a separate discussion about restricting the inheritance of environment variables. I intentionally didn't make it too restrictive in my implementation to avoid more breaking changes than necessary, but I also feel that this is not ideal from security/isolation perspective.

So I'm thinking this could be considered as a reasonable breaking change for 1.0 - would you like me to open a separate issue for that?

**Comment by caarlos0** [comment]
> So I'm thinking this could be considered as a reasonable breaking change for 1.0 - would you like me to open a separate issue for that?

yes please, I'm also a bit concerned on how it end up... pretty messy... there is even a PR opened for like a month which I'm not sure is a good idea merging :(

**Comment by caarlos0** [comment]
created #1443 

will close this one.

thaanks 🙏
